src/main/java/appeng/util/item/ItemList.java

```diff
@@ -19,126 +19,85 @@
 package appeng.util.item;
 
 
-import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.IdentityHashMap;
 import java.util.Iterator;
 import java.util.LinkedList;
-import java.util.List;
+import java.util.Map;
 import java.util.NavigableMap;
 import java.util.concurrent.ConcurrentSkipListMap;
 
-import com.google.common.collect.Lists;
-
+import net.minecraft.item.Item;
 import net.minecraftforge.oredict.OreDictionary;
 
 import appeng.api.config.FuzzyMode;
-import appeng.api.storage.data.IAEFluidStack;
 import appeng.api.storage.data.IAEItemStack;
-import appeng.api.storage.data.IAEStack;
 import appeng.api.storage.data.IItemList;
 
 
-public final class ItemList<StackType extends IAEStack> implements IItemList<StackType>
+public final class ItemList implements IItemList<IAEItemStack>
 {
 
-	private final NavigableMap<StackType, StackType> records = new ConcurrentSkipListMap<StackType, StackType>();
-	private final Class<? extends IAEStack> clz;
-
-	// private int currentPriority = Integer.MIN_VALUE;
-	public Throwable stacktrace;
-	int iteration = Integer.MIN_VALUE;
-
-	public ItemList( Class<? extends IAEStack> cla )
-	{
-		this.clz = cla;
-	}
+	private final Map<Item, NavigableMap<IAEItemStack, IAEItemStack>> records = new IdentityHashMap<Item, NavigableMap<IAEItemStack, IAEItemStack>>();
 
 	@Override
-	public synchronized void add( StackType option )
+	public void add( IAEItemStack option )
 	{
-		if( this.checkStackType( option ) )
+		if( this.isStackNull( option ) )
 		{
 			return;
 		}
 
-		StackType st = this.records.get( option );
+		final IAEItemStack st = this.getItemRecord( option.getItem() ).get( option );
 
 		if( st != null )
 		{
-			// st.setPriority( currentPriority );
 			st.add( option );
 			return;
 		}
 
-		StackType opt = (StackType) option.copy();
-		// opt.setPriority( currentPriority );
-		this.records.put( opt, opt );
-	}
-
-	private boolean checkStackType( StackType st )
-	{
-		if( st == null )
-		{
-			return true;
-		}
-
-		if( !this.clz.isInstance( st ) )
-		{
-			throw new IllegalArgumentException( "WRONG TYPE - got " + st.getClass().getName() + " expected " + this.clz.getName() );
-		}
+		IAEItemStack opt = option.copy();
 
-		return false;
+		this.putItemRecord( opt );
 	}
 
 	@Override
-	public synchronized StackType findPrecise( StackType i )
+	public IAEItemStack findPrecise( IAEItemStack i )
 	{
-		if( this.checkStackType( i ) )
+		if( this.isStackNull( i ) )
 		{
 			return null;
 		}
 
-		StackType is = this.records.get( i );
-		if( is != null )
-		{
-			return is;
-		}
+		final IAEItemStack is = this.getItemRecord( i.getItem() ).get( i );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "is". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1488)